from scripts.yaml_io import read_yaml
from scripts.extract_gb_info import get_locus
from scripts.import_user_data import Data

from scripts.get_software_parameters import (
    get_nanofilt_parameters,
    mask_regions_parameters,
    get_trimmomatic_parameters,
    get_snippy_parameters,
)


workdir: "../results/"


software_parameters = read_yaml("../user_metadata/parameters.yaml")

sample_data: Data = Data("../user_metadata/sample_data.csv")
(
    paired_illumina,
    single_illumina,
    ont_samples,
    sample_info_dic,
) = sample_data.get_options()

concatenated_illumina = list({**paired_illumina, **single_illumina}.keys())

user_metadata = read_yaml("../user_metadata/sample_data.yaml")
PROJECT_NAME = user_metadata["project_name"]
REFERENCE_FASTA = user_metadata["fasta_reference"]
REFERENCE_GB = user_metadata["gb_reference"]
REFERENCE_NAME = re.findall("(?<=references/)(.*?)(?=.fasta)", REFERENCE_FASTA)[0]
SEGMENTS = get_locus(REFERENCE_GB)

ASSEMBLER = ["snippy", "iVar"]

print(os.getcwd() + f"/../data/influenza_01_1.trimmed.fastq.gz")


include: "rules/snippy.smk"
include: "rules/iVar.smk"
include: "rules/consensus.smk"
include: "rules/gather_consensus.smk"


rule all:
    input:
        expand(
            "align_samples/{sample}/{illumina_genome_assembly_software}/{sample}_consensus.fasta",
            illumina_genome_assembly_software=ASSEMBLER,
            sample=single_illumina.keys(),
        ),
        expand(
            "align_samples/{sample}/{illumina_genome_assembly_software}/{sample}_consensus.fasta",
            illumina_genome_assembly_software=ASSEMBLER,
            sample=paired_illumina.keys(),
        ),
        expand(
            "{project}/{assembler}_consensus.fasta",
            project=PROJECT_NAME,
            assembler=["snippy", "iVar"],
        ),
        expand("{project}/compare_consensus.csv", project=PROJECT_NAME),
